### DISCUSSION TITLE
cannot find Text.Pandoc.JSON (needed for compilation ghc --make stripmath.hs)

### DISCUSSION THREAD

**Discussion by ouboub** [jgm/pandoc#10259] [General]
Hi
I asked this question years ago (and used it in Ubuntu 16.04 and the corresponding pandoc version). I am now on Ubuntu 24.04
using `pandoc 3.1.3`. I want to strip a latex file: all equations should be replaced by the word formula.
so the following was proposed to me by the pandoc author
```
import Text.Pandoc.JSON
    
main = toJSONFilter stripmath
    where stripmath (Math _ _) = Str "FORMULA"
          stripmath x = x
```
or even
```
import Text.Pandoc.JSON
    
main = toJSONFilter stripmath
    where stripmath (Math _ _) = Str "FORMULA"
          stripmath x = x
go (Math _ _) = Str "FORMULA"
go (Span (_,_,[("label",_)]) _) = Str "LABEL"
go x = x
```
both `hs` files should be compiled with 
```
ghc --make stripmath-ref.hs
ghc --make stripmath.hs
```
however in my case this returns 
```
import Text.Pandoc.JSON
    
main = toJSONFilter stripmath
    where stripmath (Math _ _) = Str "FORMULA"
          stripmath x = x
go (Math _ _) = Str "FORMULA"
go (Span (_,_,[("label",_)]) _) = Str "LABEL"
go x = x
```
I cannot find file i my Ubuntu distribution. I also pulled the github repository and still cannot find this file
What do I miss

Uwe Brauer

**Comment by ouboub** [comment]
>>> "UB" == Uwe Brauer ***@***.***> writes:

> Hi
> I asked this question years ago (and used it in Ubuntu 16.04 and the corresponding pandoc version). I am now on Ubuntu 24.04
> using `pandoc 3.1.3`. I want to strip a latex file: all equations should be replaced by the word formula.
> so the following was proposed to me by the pandoc author
> ```
> import Text.Pandoc.JSON
    
> main = toJSONFilter stripmath
>     where stripmath (Math _ _) = Str "FORMULA"
>           stripmath x = x
> ```
> or even
> ```
> import Text.Pandoc.JSON
    
> main = toJSONFilter stripmath
>     where stripmath (Math _ _) = Str "FORMULA"
>           stripmath x = x
> go (Math _ _) = Str "FORMULA"
> go (Span (_,_,[("label",_)]) _) = Str "LABEL"
> go x = x
> ```
> both `hs` files should be compiled with 
> ```
> ghc --make stripmath-ref.hs
> ghc --make stripmath.hs
> ```
> however in my case this returns 
> ```
> import Text.Pandoc.JSON
    
> main = toJSONFilter stripmath
>     where stripmath (Math _ _) = Str "FORMULA"
>           stripmath x = x
> go (Math _ _) = Str "FORMULA"
> go (Span (_,_,[("label",_)]) _) = Str "LABEL"
> go x = x
> ```
> I cannot find file i my Ubuntu distribution. I also pulled the github repository and still cannot find this file
> What do I miss

I installed via
sudo apt-get install libghc-pandoc-types-dev

Which seemed to have helped. However then the compilation gave the
following error
```
[1 of 2] Compiling Main             ( stripmath.hs, stripmath.o )

stripmath.hs:4:38: error:
    • Couldn't match type ‘[Char]’ with ‘Data.Text.Internal.Text’
      Expected: Data.Text.Internal.Text
        Actual: String
    • In the first argument of ‘Str’, namely ‘"FORMULA"’
      In the expression: Str "FORMULA"
      In an equation for ‘stripmath’:
          stripmath (Math _ _) = Str "FORMULA"
  |
4 |     where stripmath (Math _ _) = Str "FORMULA"
  |                                      ^^^^^^^^^
```

So I have still problems

**Reply by jgm** [reply]
Add this to the top of your filter:
```hs
{-# LANGUAGE OverloadedStrings #-}
```

**Reply by ouboub** [reply]
Thanks, now it compiles, however when I run
```
stripmath rev-4701272.tex rev.txt
```

the program just hangs without any result.

**Reply by jgm** [reply]
This is a filter; you use it by running pandoc and specify `--filter stripmath`

**Reply by ouboub** [reply]
Ah, of course sorry, and the most efficient way seems to do this
```
pandoc --filter /home/oub/scripts/stripmath-ref rev-4701272.tex | pandoc -t plain -o new.txt
```
BTW all the stack/haskell/cabal installation seem not necessary?!

**Reply by jgm** [reply]
You don't need to precompile the filter, but you do need pandoc-types to be installed.

By the way you can simplify, you don't need a pipe:

```
pandoc --filter YOURFILTER INPUT.tex -t plain -o OUTPUT.txt
```

**Comment by ouboub** [comment]
I tried to install pandoc-types via cabal (and stack).
So first  followed the instructions  [ghcup](https://www.haskell.org/ghcup/)
which installed stack and cabal
but 
```
cabal install --lib pandoc-types
```
leads to 
```
cabal install --lib pandoc-types
Error: cabal: Packages requested to install already exist in the environment file
at /home/oub/.ghc/x86_64-linux-9.4.7/environments/default. Overwriting them
may break other packages. Use --force-reinstalls to proceed anyway. Packages:
pandoc-types
```
the force option does not do anything, so I am puzzled

**Comment by jgm** [comment]
PS. It's now much more convenient to use a Lua filter for this.  See the documentation on the pandoc website under "Lua filters."  Writing a Lua filter to reproduce what you're doing above should be a fairly simple task.
